alphagov/paas-cf
# [#154249348] Remove IAM certificates for apps & system domains

[#154249348 Use ACM for public-facing TLS certs](https://www.pivotaltracker.com/story/show/154249348)

What?
-----

We want to replace the generated&uploaded IAM certificate with ACM
certificates from AWS.

In alphagov/paas-bootstrap#129 and alphagov/paas-cf#1202 we added logic
to create and use the new ACM certificates.

In this PR we:

 - Add temporary commit that would automatically delete the old IAM certificates. This commit shall be reverted in a follow up story.

 - Modify the CDN cloudfront distributions (productpage, etc) to use the new ACM certs.

Dependencies
------------

 1. Review and merge alphagov/paas-bootstrap#129 and alphagov/paas-cf#1202 first
 2. **REBASE** this PR after alphagov/paas-cf#1202  has been merged.
 3. They must be deployed in prod and the CDN applied in prod before merging this.

Known issues
-----------------

Our IAM policies are wrong, and we can not really run the CDN terraform completely locally or in concourse:

Locally it has problems checking the ACM certificate:

```
* module.cloudfront_paas_docs.aws_cloudfront_distribution.cdn_instance: 1 error(s) occurred:

* aws_cloudfront_distribution.cdn_instance: InvalidViewerCertificate: The specified SSL certificate doesn't exist, isn't in us-east-1 region, isn't valid, or doesn't include a valid certificate chain.
	status code: 400, request id: e1fbe566-0775-11e8-8f2d-bf0c9876e506
```

In concourse, it fails configuring the API gateway:

```
* module.redirect_paas_product_page.aws_api_gateway_rest_api.redirect_api_gw: 1 error(s) occurred:

* module.redirect_paas_product_page.aws_api_gateway_rest_api.redirect_api_gw: aws_api_gateway_rest_api.redirect_api_gw: AccessDeniedException: User: arn:aws:sts::595665891067:assumed-role/deployer-concourse/i-09c6262a5ed88ce5b is not authorized to perform: apigateway:GET on resource: arn:aws:apigateway:eu-west-1::/restapis/u9e6moaw1d
	status code: 403, request id: 9164a491-0775-11e8-87d3-af1cf64e1838
```

How to review
-------------

To test the CDN config:

 1. Checkout the previous to last commit: `git checkout HEAD~` to use the original config
 2. Create the CDN using IAM certs. Note: there are issues in our IAM policies, so you might need to run it locally and in concourse: `make dev setup_cdn_instances DEPLOY_ENV=hector ACTION=apply` and `make dev setup_cdn_instances DEPLOY_ENV=hector ACTION=apply RUN_ON_CONCOURSE=true`
 3. Checkout the latest commit: `git checkout -`
 4. Apply the config again: `make dev setup_cdn_instances DEPLOY_ENV=hector ACTION=apply` and `make dev setup_cdn_instances DEPLOY_ENV=hector ACTION=apply RUN_ON_CONCOURSE=true`
 3. Checkout the latest commit: `git checkout -`

The delete IAM cert commit:

 1. Run the pipeline
 2. cf-terraform job should run and finish.
 3. reruning cf-terraform works
 3. Check that the certificates have been deleted:

 ```
 DEPLOY_ENV=hector
 aws iam list-server-certificates | grep ${DEPLOY_ENV}
 aws s3 cp s3://gds-paas-${DEPLOY_ENV}-state/cf-certs.tfstate - # it shall fail
 aws s3 cp s3://gds-paas-${DEPLOY_ENV}-state/cf-certs.tar.gz  - | tar -tvzf -  # It shall not include apps or system
 ```

You can try the code after reverting the temporary commit that deletes the old certs:
  1. Create a tmp branch, revert the last commit and push
  2. Rerun the pipeline.


Follow up
---------

 1. IMPORTANT: Before merging, and after alphagov/paas-bootstrap#129 and alphagov/paas-cf#1202  have been applied in prod, we must apply the CDN changes
 2. After running the cleanup commit in all envs, revert it.

Who?
----

Anyone but @keymon
